### ISSUE TITLE
ACK Frames should be in Increasing Packet Number Order

### ISSUE THREAD

**Issue by nibanks** [quicwg/base-drafts#969]
The ACK frame consists of blocks in decreasing packet number order and, now that they use variable-length integers, processing these ACK blocks in reverse order is no longer feasible.

Generally, I feel that ACK blocks should be processed in increasing packet number order (start with smallest packet number), for a couple of reasons:

1. Assuming your list of Outstanding Sent Packets is ordered by increasing packet number (oldest packet at the front), for any ACK block you process, you'd expect the smallest packet number in the block to be at or near the front of the Sent Packets list.

2. If the smallest packet number isn't at the front of the Sent Packets list, then it's likely FACK/RACK could kick in, causing the front of the list to be removed immediately, further optimizing the search for the next smallest packet number that was acknowledged.

I can't see any reason for processing ACKs in the current order the frame contains them. Can anyone shed some light on why this order was originally chosen?

**Comment by mikkelfj** [comment]
There was a lengthy discussion on this and some consensus on the need of redesigning the ACK frame, but I can't seem to find the issue where this was covered.

**Comment by nibanks** [comment]
Well, if folks agree with this, it would be easy enough to change the existing ACK frame text. I'd be happy to make the changes. Find/replace in the ACK section:

- 'largest' => 'smallest'
- 'descending' => 'ascending'
- 'preceding' => 'following'

No wire format needs to change as far as I can tell.

**Comment by ianswett** [comment]
In most cases, it doesn't matter whether it's in ascending or descending order.  If one was using a linked list to store packets, then you would have to skip over a CWND of packets with the current decreasing approach, but that's not a data structure I'd recommend anyway, because it has poor cache locality.  A deque is a fairly good choice.

The original reason it was chosen is that when sending an ack frame, if there wasn't room for all the ack ranges or you didn't want to send all of them for some other reason, it's best to drop the lowest ranges, which are typically the least recently received packets, rather than the highest.  This can be accomplished by doing a first pass through the ack frame and tracking how large it'll serialize to before serializing it, but that's not ideal.

As far as I know, the serialization case is the only case where the order ends up mattering much in either computational complexity/cost or programming complexity.

I don't think I understand your second concern.

**Comment by mikkelfj** [comment]
For decoding I find the order impractical because packets are sorted by packet number oldest first. If the ACK frame had same order, both lists could be walked directly. In praxis it is not that simple because the entire packet must be validated before being processed and because there could be multiple ACK frames in a single packet so one cannot conclude timeout before seeing all.

**Comment by nibanks** [comment]
@ianswett Can you elaborate on why you think the order doesn't matter when you process the ACK packet? I am envisioning the following pseudo function:

```
process_ack_frame (ack_frame) {
  foreach (ack_block in ack_frame) {
   foreach (pkt_num in ack_block) {
      if ((sent_pkt = find_outstanding_pkt(pkt_num)) != null) {
        on_pkt_acked(sent_pkt)
      } 
    }
  }
}

find_outstanding_pkt (pkt_num) {
  foreach (pkt in outstanding_pkts) {
    if (pkt.num == pkt_num) {
      outstanding_pkts.remove(pkt)
      return pkt
    }
    if (pkt.num > pkt_num) break
  }
  return null
}
```

`process_ack_frame` is most efficient if the ACK frame/blocks are in increasing packet order. Ideally, the find function just pops the head of the list off.

**Comment by siyengar** [comment]
I think the newest first should be the order. The reason is that QUIC prefers to ACK newer packets over older packets. If we do oldest first, then we would have to do one pass over the packet to calculate the space that we have before we decide which is the oldest packet we can actually encode in the ack. This is harder to do that having the decoder just use a reverse iterator. For this reason I support the current newest first encoding

**Comment by nibanks** [comment]
@siyengar The use of variable-length integers prevents reverse traversal over the raw ACK frame.

**Comment by siyengar** [comment]
oh i see what you want to do, you want to handle acks while parsing the frame, ya in that case you can't have a reverse iterator. We parse the packet and then handle the acks which is why we can have a reverse iterator. I can see that it's shitty either way, ascending or descending. It may be less shitty as descending :)

**Comment by marten-seemann** [comment]
What I like the most about the descending order is that you can just stop writing new ranges, when you notice that the ACK is getting too big.
Furthermore, descending order saves a few bytes on the wire, since for all ranges, you're only encoding packet number differences. For ascending order, you'll need multiple bytes for the first block. It's just a couple of bytes per ACK, but since ACKs are sent frequently, this might something worth considering.

**Comment by mikkelfj** [comment]
It would be good to avoid temp allocation beyond a fixed stack allocation during decoding. There is no longer a 256 entry limit.

**Comment by nibanks** [comment]
@marten-seemann I don't understand why the ACK frame would be a different size with ascending order? It should be exactly the same, encoding gaps and blocks as packet number differences. For ascending you add the differences as you traverse ACK frame; for descending you subtract.

Also, on the issue of encoding ACK blocks in ascending order, shouldn't it be fairly to traverse you local list of packet number ranges to ACK to calculate how many of them will fit in your packet. Once that is determined, you just start from the smallest of that subset to encode the ACK frame. You can also have some simple heuristics to know (for the simple cases) if all your local ranges would fit in the packet, even if they used the full variable-length encoding for each parameter, so you don't have to calculate each individually.

**Comment by nibanks** [comment]
@marten-seemann If we changed the ACK frame to ascending order, `Largest Acknowledged` would become `Smallest Acknowledged` and all the deltas would reverse direction. You wouldn't encode anything twice.

As for the DoS, I agree, the simplistic pseudo code does have issues. Instead you would `find` the start and the end of the range, and return the set of all packets in it. But generally, the same issues apply.

**Comment by huitema** [comment]
I like one part of the descending order: immediately obtaining the "largest seen" packet number. If we agree that seeing this "largest seen" is important, then the descending order makes sense. Everything else can be solved with appropriate data structures.

**Comment by mikkelfj** [comment]
But then it must be required that highest / first ACK range really is the largest seen. There is currently no such requirement AFAIK.

**Comment by nibanks** [comment]
Why does it really matter if 'largest acknowledged' (even if it is required to be 'largest seen') be in the ACK frame's header and not just in the last ACK block? You process the entire ACK frame/blocks at essentially the same moment in time (as far as the network is concerned).

And I still I haven't heard a solution to my original problem. Given that set of outstanding sent packets is sorted in ascending order, if the ACK blocks are in descending order (and cannot be traversed in reverse order without first copying to a separate data structure and memory) how are you supposed to efficiently match ACK blocks to outstanding send packets?

The best I can think of requires a search for 'largest acknowledged' and then iteratively (per ACK block) search backwards from that point, effectively O(2n) to find all the ACK blocks; where as if the ACK blocks were in ascending order, the search is only O(n).

For example:

Outstanding Sent Packets: `{1,2,3,....1000}`
ACK Blocks: `{500 - 400},{390-300},{290-1}`

Again, assuming you don't allocate additional memory for an intermediate data structure, processing these ACK blocks in descending order would first require a search through half the Outstanding Sent Packets set to find the largest and then the same part of the set would be walked backwards for the rest of the ACK blocks. But if they were in reverse order that wouldn't be the case.

**Comment by martinduke** [comment]
I concur with all the arguments in favor of descending order.

@nibanks I see that in the general case, largest_acked can have the annoying property of being somewhere in the middle of your packet list. However, I'll note that a single block can easily be processed lowest to highest, which will tend to address your problem because...

1) in the lossless case, there is only one block anyway.
2) If the ACK only advertises one *new* gap, the earlier blocks have presumably been wiped out from your packet queue anyway.

I confess that the situation where there are large numbers of lost packets to travers in the queue could be suboptimal. One can solve this by keeping lost packets in a separate queue, or keeping a pointer to the first packet not declared lost.

Ascending order really shines when an ACK frame reports multiple *new* packet gaps. I'm not sure how to trade that off against the other considerations.

**Comment by nibanks** [comment]
@martinduke both of your reasons for saying that a block can be process easily from lowest to highest assumes little to no packet loss. We cannot make that assumption. We have to design for the most common case, which would assume packet loss. Also, in both your scenarios, ascending order processing time would be equal to descending order; but those are the **only** cases where that would be the case. In **all other cases** processing in descending order is a more expensive operation than processing in ascending order.

**Comment by martinduke** [comment]
I don't know that heavy, non-bursty packet loss is the common case.

Anyway, the increased processing expense is offset by lower cost generation of ACK frames. With ascending, I have to figure out the full size of the acknowledgment, start deleting blocks from the top if oversize, and then write into the packet.

This is just about as complex as, on processing, reading all the information in the ACK into memory and then deleting packets from the queue. If you take that approach, it's trivial to start from the lowest number. Indeed, it would be straightforward to execute this approach only if block_count is large.

To summarize:
Descending: Low-cost generation; high-cost processing under the conditions where ACKs report multiple new gaps .

Ascending: High-cost generation, low-cost processing.

**Comment by ianswett** [comment]
I think there are ways to make the processing low cost for descending as well.

I'm working on an implementation that I believe would have identical performance in both directions.  It's sort of on the back burner, but I can try to finish it soon so it's available in Chromium for people to take a look at?

**Comment by nibanks** [comment]
@ianswett I look forward to your proposal.

**Comment by siyengar** [comment]
Did some quic (sorry for the pun) back of the envelope calculations on the 2 different approaches if you want to iterate while parsing, which I think is not really necessary.

Descending order algorithm:

```
let outstanding -> rb tree, deque / vector of packets in ascending order of packetnum
let m -> num of frames in ack frame
let n -> num of outstanding packets

auto matchEnd = outstanding.end;
for (all frame in ackframe descending order) {
  auto matchedIt = search_for_first_packet_not_less_than(frame.start, outstanding.start, outstanding.end); 
  if (matchedIt == end) {
    continue;
  }
  matchEnd = matchedIt;
  while (matchedIt != outstanding.end) {
    if (matchedIt.start >= frame.start and matchedIt.end <= frame.end) {
	   ack();
	} else {
	  break;
	}
	matchedIt++;
  }
}
```
assuming the search complexity is log n

the worst complexity would probably be:
```
(m  * log n ) | (min(m, n) + min(m, n) log n)

(1) If m >> n, then an attacker would try to make all the blocks not match in which case you would get m log n
(2) If n >> m then an attacker would have greatest advantage by making only 1 block match per frame, to get you do the max number of search operations, however it is still bounded by m or n

However in case (2) the attacker could not exceed m.
so the complexity is actually

m * log n
```

The alternative approach of ascending is 

```
matchStart = outstanding.start
for (all frame in ascending order) {
  auto matchedIt = search_for_first_packet_not_less_than(frame.start, matchStart, outstanding.end); 
  if (matchedIt == end) {
    break;
  }
  matchStart = matchedIt;
  while (matchedIt != outstanding.end) {
    if (matchedIt.start >= frame.start and matchedIt.end <= frame.end) {
	   ack();
	} else {
	  break;
	}
	matchedIt++;
  }
}
```

This seems to be the same worst case complexity, so both approaches have similar DoS capabilities however the best case complexity is better since you can exit early with ascending

Practically though I don't think this is a big deal implementations will have multiple layers in the code, one which parses the data into an in memory frame structure after which you can use either an ascending or descending algorithm. 

For example even with an ascending data structure in the ACK frame you still need to incur high cost again, because you need to know the largest acked packet for loss detection in the case of time based loss detection as well as early retransmit timers which involves iterating over the outstanding packets again, which is much better to do if you can iterate over the frames just once which you can do if you pre-parse the entire data structure into memory.

I would rather make it easier to serialize data into the packet so that senders of ACKs will do the right thing of prioritizing ACKing larger packets over optimizing streaming parsing.

**Comment by mikkelfj** [comment]
Good input.

Some concerns: assuming you have descending order in frame. If you can seek to the end then process in ascending order, this is probably fine. But if you cannot, you either rely on assumptions about log N handling, or you must allocate a temporary data structure. Earlier there was a limit on 256 entries but this limit is gone. The sender knowns how many entries it creates and can easily keep number of entries below a limit. The receiver can't. Especially if it must expect very large MTU.

The receiver also do not control the frequence of ACK frames and hence the amount of resources to dedicate while the sender does. The sender is also the one that can do DoS.

As to the log N assumption, that doesn't work if you use a bitmap scan. A bitmap scan is resistant to DoS because it is O(N). Since packets now are in incremental order, this is a practical approach. 

Scanning the bitmap from most recent to older probably has a number of problems that I currently do not recall. Generally you want to work with the oldest packet first.

I can see the idea of building in descending order but you can do two passes to compute the size and then build ascending.

**Comment by mikkelfj** [comment]
There also ought to be a requirement that if there are multiple ACK frames in a packet, they must be in ascending order, which of course puts even more strain on the sender, but it is for the same reason. It is simpler to require that there is at most one ACK frame per packet.

**Comment by janaiyengar** [comment]
I don't see a reason at the moment to have two ACK frames per packet (especially with the varint representations). That said, I don't see this perf issue as a reason to require it... we can certainly recommend it strongly (SHOULD).

**Comment by mikkelfj** [comment]
If there is a MUST, the receiver can have an implementation that takes an ACK frame, makes conclusion on age and time, and do updates. With SHOULD, it needs to add additional complexity to manage information from both frames, even if a second frame is rare. Granted, the ACK frame must be parced once to validate the packet, but it is still simpler with at most one.

**Comment by janaiyengar** [comment]
The receiver doesn't need to do any ack info aggregation per packet. Just treat it like two separate ACK frames were received.

**Comment by mikkelfj** [comment]
OK, that simplifies things.

**Comment by nibanks** [comment]
@siyengar Thanks for the write up. I can agree with most of what you wrote except: 

> Practically though I don't think this is a big deal implementations will have multiple layers in the code, one which parses the data into an in memory frame structure after which you can use either an ascending or descending algorithm.

There is no reason to require an implementation to use any structure other than the wire format. Why can't it process the packet directly?

As far as the complexity of each design, I agree they can essentially have the same worst case, but then our decision should be based on the common case. Generally, ACKs will be received for the oldest outstanding packets, correct? And if the oldest packet number ACKed in an ACK frame is significantly newer than your oldest outstanding packet, then you will be throwing (mark as lost) that old packet out anyways.

If you use ascending order then you could rely exclusively on a linear walk of your outstanding packets, starting at the beginning. For each packet in your list the packet, that is less than the min of the current ACK block, it will either be ACKed or lost (for the most part). Once you get to the last packet number being ACKed in the frame, you stop walking the list.

Your worst case then is a single ACK block that ACKs your most recent packet; but in that case, it would likely cause you to throw away all your older packets anyways, which require you to walk the entire list.

**Comment by siyengar** [comment]
@nibanks I agree with you, there is no reason for an implementation to require any structure other than a wire format. For example the for loop could just be an iterator adapter in C++ that lazily processes packets. There are just tradeoffs with both approaches and I was just trying to point out the tradeoffs.

>  Generally, ACKs will be received for the oldest outstanding packets, correct?

In the common case the ACK  would be received for the most recently received packet first. Quic's loss detection algorithms rely on the sender prioritizing ACKing higher packets first for example for stuff like TLP or RTO to work correctly. A descending first algorithm kind of makes this feel more natural to to serialize packets correctly. 

Even if you use ascending order with streaming parsing you'd have to probably iterate over the packets again anyway after you're done processing the acks. There are 2 types of loss detection in QUIC. For timer based loss detection you'd have to know what the largest acked packet is before triggering it on all the packets. Even for packet order based loss detection you probably need to iterate over most of the packets in the common case because there is no guarantee given acked packet n, packets below it are less than n - kReorderingThreshold, I haven't worked out the algorithmic details for that yet and we might be able to make this more efficient by caching the minOutstandingPacketNumber, but I don't think that will change the complexity analysis.

**Comment by nibanks** [comment]
@siyengar I agree that ACKs will be sent for the most recently received packets. But I am assuming that most of the time, all received packets will be able to be ACKed; that most of the time, space won't be an issue and all the ACK blocks will fit. There might be a couple gaps on average (it would be great if someone had some numbers related to the size and consistency of ACK blocks with certain network conditions).

So if we assume that generally, all packets received are acknowledged then we are back to my original argument. The oldest packet number in the ACK frame will generally be your oldest outstanding packet (or close to it). So if you process both sets, ACK blocks and outstanding packets, together, in ascending order, you will be able to do a single pass of both and determine if an outstanding packet is ACKed, Lost, or Still Pending. And at most, you will only process your outstanding packet list up to the largest ACKed packet in the ACK frame.

**Comment by nibanks** [comment]
@ianswett Any updates? I still don't see a way that descending order could ever be as good as ascending order. I'm still thinking of this example where you have a couple of gaps of lost packets.

Outstanding Sent Packets: `{1,2,3,....1000}`
ACK Blocks: `{500 - 400},{390-300},{290-1}`

Assuming the final result is that packets `{1-290}, {300-390}, {400-500}` are acknowledged and packets `{291-299}, {391-399}` are marked as lost (FACK), that means you would have to traverse the entire first 500 packets of the outstanding list. If you walk the list in ascending order, you only touch each packet once, and no more packets than necessary. But to do that (without an extra preprocessing step and associated temporary memory) you would need the ACK blocks in ascending order.

**Comment by ianswett** [comment]
No updates, but I'll explain what my approach is.

I store an IntervalSet of the previously acknowledged packets from the lowest one I'm tracking to the largest previously acknowledged.  When I read the ack frame, I'm incrementally comparing the intervals I'm parsing with the IntervalSet of previously acknowledged packets and storing any newly acknowledged packets in a list and adding them to the IntervalSet for the future.  Then I do a second pass over the full data structure with the newly acknowledged packets in the list.  The list isn't necessary, so you could do it in one pass, but it fits into my current implementation better, so it's what I'm doing for now.  My congestion control APIs take a list of newly acked packets and their size in bytes, so I have to create that list anyway.

You have to ensure you don't allow someone to acknowledge packets larger than the largest ever sent, or you can allocate a ton of extra temporary memory(a perk of doing this without a temporary list), but otherwise, this should turn out to be slightly faster for decreasing order, since I can ignore any extra ack ranges that are below the smallest packet I'm still waiting for an acknowledgement for.

**Comment by nibanks** [comment]
Thanks @ianswett for the detailed explanation. Do you use this IntervalSet of previously acknowledged packets for anything other than this purpose? Otherwise, to me this still seems just like an intermediate data structure for processing the ACK block. We haven't had need of a similar object in our code and I would be hesitant to add it just for this.

My ideal goal is still:

1. Process ACK block directly from the wire format.
2. Require no extra processing steps or temporary storage.
3. Process the necessary outstanding packet structures only.

I see a solution with ascending order but descending order breaks at least one of the above goals.

**Comment by ianswett** [comment]
It has two purposes.  One is this, and the other is that reusing the existing ack frame trades a small amount of memory for a large-ish reduction in memory allocations.  It might even be possible for it to be a net savings of memory allocated in the typical case, by removing the redundant information of what has already been acknowledged from the more complete data structure, but I haven't tried that yet.

So this gives you 1, 2, and 3 with the only cost being complexity of an optimized data structure to speed part 3, which happens to look exactly like how you likely store the QUIC ack frame you're putting received packets into and sending.  It does assume you're using a deque or something similar that makes it fast to seek to a given entry, but that's the best style I've found in practice.

How are you intending to store and find newly acked packets?  It's worth keeping in mind that after loss, you'll likely want to track packets that have already been retransmitted for a while, in case they were spuriously retransmitted, so you may commonly end up skipping over those on the way up.  It's also worth keeping in mind that linked lists have awful performance from a CPU cache perspective.

**Comment by mikkelfj** [comment]
Given that packet numbers are in incremental order it can be done much more efficiently in time and space by scanning chained blocks of bitmaps. If there is a large gap you just link two bitmaps. Doing this efficiently requires ack's in increasing order. If your chain grows too long, you are subject to abuse, or conditions too bad to maintain a connection.

The initial descend into the bitmap can be done by remembering the previous start location, or by using an efficient index into each bitmap block.

If you arbitrarily loose every second packet, an interval map is using a lot of memory in comparison, not to mention a log2 lookup.

So, it would be rather sad to have to allocate a lot of temporary memory for no good reason.

**Comment by mikkelfj** [comment]
Well, maybe chained bitmaps could be revised to be scanned in the opposite direction but you likely need a second pass to make retransmission decisions.

**Comment by ianswett** [comment]
I haven't thought through the details of using bitmaps as much, except to say that real-world packet loss works well with ranges, even if bitmaps have better worst case properties.  I also find it convenient that ranges match up with the ack format nicely.  If packet loss was random, GQUIC would likely have used bitmaps for the ack format.

Marking packets as lost is currently a separate pass in the GQUIC code, because initially we were using pluggable loss detection.  Doing descending order works fine for detecting lost packets, and it doesn't matter which order packets are lost in to make retransmission decisions, because we can decide what to send after all lost packets are determined.

If one did use a linked list without any sort of side data structure, I don't think it matters which direction you iterate in.  Either way, you're going to have to check all packets being tracked in the list between the largest acked packet and the smallest.  I don't think the iteration order matters?  You would have to store an extra pointer to quickly access the packet larger than the largest previously acknowledged to speed access to largest acked, but that seems simple.

**Comment by nibanks** [comment]
If you store a linked list without any side structure, and the largest packet number is the first value you process, you have to process the items from the beginning to the largest twice. Once to find the actual largest item, and then another as you walk backward to process the ACK blocks.

**Comment by mikkelfj** [comment]
@nibanks you just reverse the list order in the first pass :) (never mind messing with the cache)

**Comment by ianswett** [comment]
The beginning being the smallest packet?  You can start at the largest, or even better, whichever is closer to largest acked(the largest sent or the largest previously acked).  It does mean touching a few extra elements, but typically those are being acknowledged and going to be accessed anyway.

That being said, in many scenarios, having a side data structure is going to make this faster.

**Comment by nibanks** [comment]
When you get an ACK frame, on average, where do you expect the packet number ranges to reside, in relation to our set of outstanding packets? Generally for the oldest outstanding packets?

Then, assuming FACK, if you do receive ACK blocks/frame for packet numbers more recent (larger) than your oldest packets, you would mark those 'skipped' packets as lost, therefore removing them from your outstanding packets set, right?

In that case, you essentially 'touch' all the oldest packets, up to the largest packet number in the ACK frame. So why not just walk both the ACK frame and your set of outstanding packets, in ascending order, together marking packets as either lost or ACK?

**Comment by mikkelfj** [comment]
Although my previous comment wasn’t entirely serious, if you do insist on using a list, and the ack order is reverse, then it makes sense to root the list at largest acked and point to older packets, then root a separate list in forward direction with a tail pointer. You move elements from the forward list to the reverse list while searching for a new largets acked and add new packets at the tail.

**Comment by ianswett** [comment]
You generally expect the next packets acknowledged to be close to the largest acked.

We've found it's important to keep around older packets for a while after they've been lost to detect spurious retransmits, which are not that infrequent.  Currently, we keep around about an RTT of old packets, in the hopes that QUIC can tolerate a full RTT of reordering in the worst case.  That's probably overkill for most cases, but ideally it only costs a bit of memory that doesn't end up being accessed, so there should be little to no CPU overhead.

But the key difference here is the data structure we're discussing using.  For a deque or anything similar to it, I think something like the approach I described works well and order doesn't matter.  If you use only a linked list, then ascending order is simpler to implement efficiently, but I think the real-world performance of either direction is very similar if you follow some of MikkelFJ's suggestions.

**Comment by nibanks** [comment]
Thanks @ianswett. Really appreciate the information. I agree that if you use a more complicated data structure to represent your outstanding packets than a list, you have the ability to process ACK blocks, independent of order, pretty much the same.

But I am still trying to understand why not use ascending order. The main argument I have heard is that, if you have too many packet numbers to ACK, then you always want to ACK the largest/most recent ones; and that is easier to do if you start at the largest and just continue to write the ACK blocks until you run out of room. I agree that it is easier. I would propose two solutions:

1. Keep track of the amount of space it would require you to encode your current set of ACK'ed packet numbers and only keep around what can fit into a single packet.

2. Have writing the ACK frame be a two step process. First to walk the your list in descending order to figure out how many packet numbers would fit in the packet. Then, starting at the smallest that would fit, work upward and encode.

Neither of those seem too complicated to me, and don't require more complicated data structures or significant extra processing time.

**Comment by ianswett** [comment]
Re 1, the issue is really if you'd like to bundle an ack with data.  If you only send acks in ack-only packets, it should be fairly easy to ensure it always fits, though the combination of varints and ranges potentially being merged makes it more complex than you'd think, if you want to stay close to the limit.  I would not advise anyone write this code, because I think it's error-prone.

Re 2, sure that can definitely be done.  And that's what I'd recommend if we did increasing order, but given processing acks in decreasing order is easy(in my case, it's equivalently costly and equivalent complexity), I think this double-pass is more complex and costly than the processing acks in decreasing order.

**Comment by nibanks** [comment]
> I think this double-pass is more complex and costly than the processing acks in decreasing order.

I would generally agree, and I did do exactly that before the varint changes. But with varint, I cannot process the ACK block in reverse order without first copying it to some other format in some other memory. Assuming worst case of a a full packet with a lot of 1 byte gaps/blocks encoded with varint, that could quickly explode if my temporary data structure uses the full  8 bytes to represent each.

**Comment by mikkelfj** [comment]
just a thought: I like the current varint format, but if varints were encoded with highbit in all but last byte, then they could be processed in both directions if you know when to stop. utf-8 is similar self-sync'ing encoding.

**Comment by marten-seemann** [comment]
> But with varint, I cannot process the ACK block in reverse order without first copying it to some other format in some other memory. Assuming worst case of a a full packet with a lot of 1 byte gaps/blocks encoded with varint

I don't think varint changes anything here. The old ACK format also allowed for 1 byte gaps / blocks.

>  that could quickly explode if my temporary data structure uses the full 8 bytes to represent each.

The thought of having a ~ 10 kb data structure to represent the ranges of a single ACK frame is unpleasant, however, if you're concerned about allocations, you only need one of those per connection, if you reuse it.

**Comment by nibanks** [comment]
@marten-seemann The temporary structure wasn't needed before varints, as I could parse the wire format in reverse. As to having the 10kb structure always allocated with the connection, I agree that is a solution, though it seems pretty wasteful, considering that worst case will almost never be hit.

**Comment by nibanks** [comment]
Also, it could get more complicated, as the max size of the temporary data structure would depend on MTU, which could change, or just outright, be very large. Again, worst case if 64KB was the MTU, then that temporary structure is HUGE.

**Comment by ianswett** [comment]
You can close the connection if any packets larger than largest acked are acknowledged and skip over any ranges less than the smallest packet you're still tracking, so I believe any portion of an ack frame you have to process is still much smaller than the datastructure of outstanding packets.

**Comment by larseggert** [comment]
I don't know where the 10KB size comes from. See https://web.engr.oregonstate.edu/~erwig/papers/Diet_JFP98.pdf which is what I use in https://github.com/NTAP/quant/blob/master/lib/src/diet.c. It uses 24B per range.

**Comment by huitema** [comment]
Note that there are two problems. The ACK list needs access both up and down, but a doubly linked list does that. The ACK generation needs a SACK dashboard similar to that of TCP, and that requires a small list of range descriptors, but that's not hard either.

I would much rather keep the spec as is. Nick is raising a performance optimization issue. Personally, I would rather like to see the spec stabilize before working full hog on performance optimization.

**Comment by mikkelfj** [comment]
The ACK format is fundamental to the implementations data-structures at a deep level, it needs to be roughly sorted out in v1. I can't imagine two completely different approaches running side by side to support v1 and v2 even if I can agree that performance details in general can be sorted later. Even if it is not that hard to run in either direction, once you choose a direction, it is hard to change unless you use a very simplistic approach.

**Comment by mikkelfj** [comment]
Here is one way the reverse list could be processed in ascending order, assuming MTU < 2^64.

- Stack allocate a 2-byte offset stack of 100 entries (or more if you like)
- Scan the ack frame by pushing an offset for 1 in every 100 range entries seen.
- If the offset table is exhausted, drop the rest of the ACK frame, or cry abuse / bad connection
- Stack allocate a range table of 100 ranges (2 x 8 bytes x 100).
- While offset stack not empty pop an offset from the offset table and fill the range table by reading up to 100 ranges then process that range back to front.

Optimisations:

- Treat the last offset (first in while loop) specially since it does not represent a fixed 100 ranges, and often is the only offset to process.
- Use recursion with max depth = 100 to implement the offset stack, but pass the prealocated range table by reference for memory reuse.
- When the ACK frame has a count of ranges in the header as opposed to a terminal null range, it is easy to detect if there are at most 100 ranges and go for the fast path immediately.

This has low memory overhead and works in L1 cache by reusing the same temp table repeatedly. Likewise, dual passes over the ACK frame is cheap from a cache perspective. Varints are reasonably easy to scan without decoding in the first pass.

While the algorithm may sound a bit involved, it is probably simpler and faster than encoding the ACK frame in ascending order.

**Comment by mikkelfj** [comment]
With the above algorithm I can sort of buy into the descending ACK argument, but that leaves me with:

The transport doc does not currently explicitly state the ACK Block Count is variable length, but I presume that it is. If so, how do you efficiently do a single pass ACK frame construction when you don't know the size of the count field? Do you just use a large enough integer encoding? Do you use memmove to retro-fit? Do you preprocess to decide?

Transport 8.15:
> ACK Block Count:
The number of Additional ACK Block (and Gap) fields after the First ACK Block.

**Comment by nibanks** [comment]
@mikkelfj an easy solution to the ACK Block Count variable is to always encode it using 2 bytes, even if the value could have used 1 byte. That way, you can go back and write the field afterwards.

**Comment by martinthomson** [comment]
It seems like @nibanks is on his own here, but rather than just close this, we will wait to get broader experience with implementation.  Once more people have real code that does this, then we will have the information we need to make this decision.  Hopefully sunk costs don't drive the outcome too much at that point.

**Comment by martinthomson** [comment]
Tokyo conclusion: @nibanks wasn't even sure that this was still open.  There is a trade-off here between optimizing for reading and optimizing for writing.  There's an implementation overhead for things like RACK/FACK when reading in reverse order.

Closing on the basis that we don't have a very good reason to change.
